### PR TITLE
Add joke data crawler

### PR DESCRIPTION
**Pull request by zeagoldgh**

### CODE REVIEW

File: api-server/build/api-server/src/main.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+joke_data = {
+    "id": data["id"],
+    "joke": data["value"],
+} 
+
+print(joke_data)
+
+with open("joke.json", "w") as f:
+        f.write(json.dumps(joke_data))
```
Comment (fabianschmauder) [inline api-server/build/api-server/src/main.py]: better move this part into a function that you can call. Otherwise it will directly be called

Comment (fabianschmauder) [reply api-server/build/api-server/src/main.py]: you create a file. where do you use it?

File: joke-data-crawler/src/s3_utils.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+s3_client = boto3.client("s3")
+
+joke_data_bucket = "joke-data-src-bucket-dzaa1417"
+
+def save_joke_to_s3(joke):
+    joke_data_string = json.dumps(joke)
+    joke_data_bytes = str.encode(joke_data_string)
+    s3_client.put_object(
+        Body = "somebody",
```
Comment (fabianschmauder) [inline joke-data-crawler/src/s3_utils.py]: some body should be your joke content

**Review by fabianschmauder — COMMENTED** [review]
please remove the zip files from your git project. 
I think the main.py is just for testing?
You do not save your joke content to s3.

Do you also have terraform files?

File: upload-src.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+sh setup-src-bucket.sh
+
+mkdir build
+
+cd joke-data-crawler/src
+zip -r ../../build/joke-crawler.zip .
+cd ../..
+
+# "upload to s3"
+aws s3 cp build/joke-data-crawler.zip s3://joke-src-bucket-dzaa1417/
```
Comment (fabianschmauder) [inline upload-src.sh]: there is a typo

use:
aws s3 cp build/joke-crawler.zip s3://joke-src-bucket-dzaa1417/

File: upload-src.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+cd requests-layer
+#pip3 install -r requirements.txt --target python/lib/python3.9/site-packages
```
Comment (fabianschmauder) [inline upload-src.sh]: maybe add the install again

File: joke-data-crawler/src/s3_get_joke.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+import json
+
+s3 = boto3.resource("s3")
+
+joke_data_bucket = ""
+
+def get_joke_data(id):
+    obj = s3.Object(joke_data_bucket, id+".json")
+    joke_string = obj.get()['Body'].read().decode('utf-8')
+    return json.load(joke_string)
```
Comment (fabianschmauder) [inline joke-data-crawler/src/s3_get_joke.py]: empty line

File: joke-data-crawler/src/s3_upload_joke.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import json
+import boto3
+
+s3_client = boto3.client("s3")
+
+def upload_joke(joke_data_to_save):
+    with open("joke.json", "w") as file:
+        file.write(json.dumps(joke_data_to_save))
+
+s3_client.upload_file('joke.json', 'joke-src-bucket-dzaa1417', 'joke.json')
```
Comment (fabianschmauder) [inline joke-data-crawler/src/s3_upload_joke.py]: empty line missing

File: joke-data-crawler/src/s3_upload_joke.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import json
+import boto3
+
+s3_client = boto3.client("s3")
```
Comment (fabianschmauder) [inline joke-data-crawler/src/s3_upload_joke.py]: do you use this for testing?

In lambda this function will not work

Comment (fabianschmauder) [reply joke-data-crawler/src/s3_upload_joke.py]: you can  only save to temp in lambda

Comment (fabianschmauder) [reply joke-data-crawler/src/s3_upload_joke.py]: upload ist directyl called in lambda

**Review by fabianschmauder — COMMENTED** [review]
please remove all files that you do not need

File: joke-data-crawler/src/s3_upload_joke.py (hunk 7 above)
Comment (zeagoldgh) [reply joke-data-crawler/src/s3_upload_joke.py]: yes just for testing. I am pushing the real codes soon

File: .gitignore [hunk 8]
```diff
@@ -1 +1,4 @@
-.DS_Store
\ No newline at end of file
+.DS_Store
+build
+*_venv
+__pycache__
```
Comment (fabianschmauder) [inline .gitignore]: missing empty line ;)

File: infrastructure/joke_crawler.tf [hunk 9]
```diff
@@ -0,0 +1,38 @@
+resource "aws_lambda_function" "joke_crawler" {
+  function_name = "joke_crawler"
+  s3_bucket     = "joke-src-bucket-dzaa1417"
+  s3_key        = "joke-crawler.zip"
+  role          = "arn:aws:iam::208631683274:role/LabRole"
```
Comment (fabianschmauder) [inline infrastructure/joke_crawler.tf]: I would use the terraform provided account id. 
Maybe define a local variable to create the role arn

File: infrastructure/joke_crawler.tf [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    arn = aws_lambda_function.joke_crawler.arn
+}
+
+resource "aws_lambda_permission" "allow_cloudwatch_to_call_joke_crawler" {
+    statement_id = "AllowExecutionFromCloudWatch"
+    action = "lambda:InvokeFunction"
+    function_name = aws_lambda_function.joke_crawler.function_name
+    principal = "events.amazonaws.com"
+    source_arn = aws_cloudwatch_event_rule.every_seven_minutes.arn
+}
```
Comment (irenepapaspyratos) [inline infrastructure/joke_crawler.tf]: Line missing

File: joke-data-crawler/src/chucknorrisapi.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import requests
+
+url = "https://api.chucknorris.io/jokes/random"
+
+
+def load_jokes():
+    response = requests.get(url)
+
+    return response.json()
```
Comment (irenepapaspyratos) [inline joke-data-crawler/src/chucknorrisapi.py]: Line missing

File: joke-data-crawler/src/crawler.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        })
+
+def handler(event, context):
+    jokes_from_api = load_jokes()
+    jokes = map_jokes(jokes_from_api)
+    save_joke_to_s3(jokes)
+    
+
+if __name__ == "__main__":
+    handler({},{})
```
Comment (irenepapaspyratos) [inline joke-data-crawler/src/crawler.py]: LINE missing

File: joke-data-crawler/src/s3_utils.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    joke_data_bytes = str.encode(joke_data_string)
+    s3_client.put_object(
+        Body = joke_data_bytes,
+        Bucket = joke_data_bucket,
+        Key = joke["id"]+".json"
+    )
+
+def save_jokes_to_s3(jokes):
+    for joke in jokes:
+        save_joke_to_s3(joke)
```
Comment (irenepapaspyratos) [inline joke-data-crawler/src/s3_utils.py]: LINE MISSING

File: requests-layer/.gitignore [hunk 14]
```diff
@@ -0,0 +1 @@
+python
```
Comment (irenepapaspyratos) [inline requests-layer/.gitignore]: LINE MISSING !!!!!

File: joke-data-crawler/src/chucknorrisapi.py [hunk 15]
```diff
@@ -0,0 +1,9 @@
+import requests
+
+url = "https://api.chucknorris.io/jokes/random"
+
+
+def load_jokes():
```
Comment (fabianschmauder) [inline joke-data-crawler/src/chucknorrisapi.py]:
```suggestion
def load_joke():
```
i would rename the function. It only returns one joke

File: setup-joke-data-bucket.sh [hunk 16]
```diff
@@ -0,0 +1,8 @@
+S3_BUCKET_NAME=joke-data-src-bucket-dzaa1417
+REGION=us-west-2
+if aws s3 ls "s3://$S3_BUCKET_NAME" 2>&1 | grep -q 'An error occurred'
+then
+    aws s3api create-bucket --bucket $S3_BUCKET_NAME --region $REGION --create-bucket-configuration LocationConstraint=$REGION
+else
+    echo "bucket already exists"
+fi
```
Comment (irenepapaspyratos) [inline setup-joke-data-bucket.sh]: LINE MISSING !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

File: setup-src-bucket.sh [hunk 17]
```diff
@@ -0,0 +1,8 @@
+S3_BUCKET_NAME=joke-src-bucket-dzaa1417
+REGION=us-west-2
+if aws s3 ls "s3://$S3_BUCKET_NAME" 2>&1 | grep -q 'An error occurred'
+then
+    aws s3api create-bucket --bucket $S3_BUCKET_NAME --region $REGION --create-bucket-configuration LocationConstraint=$REGION
+else
+    echo "bucket already exists"
+fi
```
Comment (irenepapaspyratos) [inline setup-src-bucket.sh]: LIIIIIIIIINE !!!

File: joke-data-crawler/src/crawler.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
+from chucknorrisapi import load_jokes
```
Comment (fabianschmauder) [inline joke-data-crawler/src/crawler.py]:
```suggestion
from chucknorrisapi import load_joke
```

File: joke-data-crawler/src/crawler.py [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+        
+    return ({
+           "id": joke_from_api["id"],
+    "joke_value": joke_from_api["value"],
+        })
+
+def handler(event, context):
+    jokes_from_api = load_jokes()
+    jokes = map_jokes(jokes_from_api)
+    save_joke_to_s3(jokes)
```
Comment (fabianschmauder) [inline joke-data-crawler/src/crawler.py]:
```suggestion
    joke_from_api = load_joke()
    joke = map_joke(joke_from_api)
    save_joke_to_s3(jokes)
```

Comment (fabianschmauder) [reply joke-data-crawler/src/crawler.py]: it is just a single joke

File: upload-src.sh [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+# "upload to s3"
+aws s3 cp build/joke-crawler.zip s3://joke-src-bucket-dzaa1417/
+
+cd requests-layer
+pip3 install -r requirements.txt --target python/lib/python3.9/site-packages
+zip -r ../build/requests-layer.zip .
+cd ..
+
+echo  "upload to s3"
+aws s3 cp build/requests-layer.zip s3://joke-src-bucket-dzaa1417/
```
Comment (irenepapaspyratos) [inline upload-src.sh]: LIIIIIIINE MISSIIIIIIIIIIING !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 🤬

File: joke-data-crawler/src/s3_utils.py (hunk 13 above)
Comment (fabianschmauder) [inline joke-data-crawler/src/s3_utils.py]: you do not use this code. Therefore I would remove it

File: requests-layer/.gitignore (hunk 14 above)
Comment (fabianschmauder) [inline requests-layer/.gitignore]: empty line

File: requests-layer/requirements.txt [hunk 21]
```diff
@@ -0,0 +1,2 @@
+requests==2.28.1
+boto3==1.26.3
```
Comment (fabianschmauder) [inline requests-layer/requirements.txt]: I think you do not need boto 3 in the layer

File: setup-joke-data-bucket.sh [hunk 22]
```diff
@@ -0,0 +1,8 @@
+S3_BUCKET_NAME=joke-data-src-bucket-dzaa1417
```
Comment (fabianschmauder) [inline setup-joke-data-bucket.sh]: joke-data-src-bucket-dzaa1417 sounds like src code.
maybe use joke-data-bucket-dzaa1417

**Review by fabianschmauder — COMMENTED** [review]
Great. Just a few suggestions

**Pull request merged**